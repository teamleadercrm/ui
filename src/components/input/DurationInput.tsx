import React, { FocusEvent, FormEvent, forwardRef, KeyboardEvent, useImperativeHandle, useRef } from 'react';
import Box from '../box';
import cx from 'classnames';
import theme from './theme.css';
import NumericInput, { NumericInputProps } from './NumericInput';
import { GenericComponent } from '../../@types/types';
import { BoxProps } from '../box/Box';
import { SIZES } from '../../constants';

const MINUTES_STEP = 15;

export interface DurationInputProps
  extends Omit<NumericInputProps, 'className' | 'id' | 'max' | 'onBlur' | 'onChange' | 'onKeyDown' | 'size' | 'value'>,
    Omit<BoxProps, 'size' | 'ref'> {
  /** Sets a class name for the wrapper to give custom styles. */
  className?: string;
  /** The id of the container containing the input field. */
  id?: string | undefined;
  /** The maximum value that can be inputted. */
  max?: number;
  /** Callback function that is fired when blurring the input field. */
  onBlur?: (event: FocusEvent<HTMLElement>) => void;
  /** Callback function that is fired when the component's value changes. */
  onChange?: (value?: { hours?: number; minutes?: number }) => void;
  /** Callback function that is fired when the keyboard is touched. */
  onKeyDown?: (event: KeyboardEvent<HTMLElement>) => void;
  /** Size of the input element. */
  size?: Exclude<typeof SIZES[number], 'tiny' | 'fullscreen' | 'smallest' | 'hero'>;
  /** Current value of the input element. */
  value?: {
    hours?: number;
    minutes?: number;
  };
}

const DurationInput: GenericComponent<DurationInputProps> = forwardRef<HTMLElement, DurationInputProps>(
  (
    {
      autoFocus,
      className,
      error,
      id,
      max,
      onBlur,
      onChange,
      onFocus,
      onKeyDown,
      size = 'medium',
      textAlignRight = false,
      value,
    }: DurationInputProps,
    forwardedRef,
  ) => {
    const containerRef = useRef<HTMLElement>(null);
    useImperativeHandle<HTMLElement | null, HTMLElement | null>(forwardedRef, () => containerRef.current);

    const handleHoursChanged = (_: FormEvent<HTMLElement>, hours: string) => {
      if (!onChange) {
        return;
      }

      if (hours === '') {
        if (typeof value?.minutes === 'undefined') {
          onChange();
          return;
        }

        onChange({
          minutes: value?.minutes,
        });
        return;
      }

      const parsedHours = parseInt(hours);
      if (!Number.isInteger(parsedHours)) {
        return;
      }

      onChange({
        ...value,
        hours: parsedHours,
      });
    };

    const handleMinutesChange = (_: FormEvent<HTMLElement>, minutes: string) => {
      if (!onChange) {
        return;
      }

      if (minutes === '') {
        if (typeof value?.hours === 'undefined') {
          onChange();
          return;
        }

        onChange({
          hours: value?.hours,
        });
        return;
      }

      const parsedMinutes = parseInt(minutes);
      if (!Number.isInteger(parsedMinutes)) {
        return;
      }

      if (parsedMinutes > 60) {
        return;
      }

      if (parsedMinutes === 60) {
        onChange({
          hours: (value?.hours || 0) + 1,
          minutes: 0,
        });
        return;
      }

      if (parsedMinutes < 0) {
        onChange({
          hours: (value?.hours || 1) - 1,
          minutes: 60 - MINUTES_STEP,
        });
        return;
      }

      onChange({
        ...value,
        minutes: parsedMinutes,
      });
    };

    /**
     * This blur handler is a special one, because this input consists of multiple focus points:
     * the two inputs themselves, and the four increment/decrement buttons
     * we filter out the blur events caused by changing focus between these, so the onBlur event only
     * gets triggered when really blurring outside of the entire component
     */
    const handleBlur = (event: FocusEvent<HTMLElement>) => {
      if (!event.relatedTarget) {
        onBlur && onBlur(event);
        return;
      }

      const inputs = containerRef.current?.getElementsByTagName('input') || [];
      if (Array.from(inputs).find((node) => node === event.relatedTarget)) {
        return;
      }

      const buttons = containerRef.current?.getElementsByTagName('button') || [];
      if (Array.from(buttons).find((node) => node === event.relatedTarget)) {
        return;
      }

      onBlur && onBlur(event);
    };

    const transformToPaddedNumber = (number: number) => {
      return number < 10 ? `0${number}` : number.toString();
    };

    const getHoursString = () => {
      const hours = value?.hours;

      if (Number.isInteger(hours)) {
        return transformToPaddedNumber(hours as number);
      } else if (hours) {
        return hours.toString();
      }

      return undefined;
    };

    const getMinutesString = () => {
      const minutes = value?.minutes;

      if (Number.isInteger(minutes)) {
        return transformToPaddedNumber(minutes as number);
      } else if (minutes) {
        return minutes.toString();
      }

      return undefined;
    };

    const getMaxHoursAndMinutes = (): { maxHours: number | undefined; maxMinutes: number | undefined } => {
      if (max) {
        return {
          maxHours: Math.floor(max / 60),
          maxMinutes: max % 60,
        };
      }

      return {
        maxHours: undefined,
        maxMinutes: undefined,
      };
    };

    const getMaxHourLength = () => {
      const { maxHours } = getMaxHoursAndMinutes();
      const hoursString = getHoursString();

      if (hoursString && maxHours) {
        let maxHourLength = maxHours.toString().length;

        // if it starts with 0, allow one more so we don't block typing (because we autopad the hours)
        if (hoursString.length >= maxHourLength && hoursString[0] === '0') {
          maxHourLength += 1;
        }

        return maxHourLength;
      }

      return undefined;
    };

    const getMaxMinuteLength = () => {
      const minutesString = getMinutesString();

      // if it starts with 0, allow one more so we don't block typing (because we autopad the minutes)
      return minutesString && minutesString.length >= 2 && minutesString[0] !== '0' ? 2 : 3;
    };

    const isMaximumMinutesLimited = () => {
      const { maxHours } = getMaxHoursAndMinutes();

      // Minutes are relative to the already filled in hours, so the max needs to be set on the fly
      return (max && value?.hours && maxHours && value.hours >= maxHours) || maxHours === 0;
    };

    return (
      <Box
        data-teamleader-ui="duration-input"
        id={id}
        display="flex"
        alignItems="center"
        ref={containerRef}
        className={cx(className, theme['duration-input-numeric-container'], { [theme['has-error']]: error })}
      >
        <NumericInput
          placeholder="00"
          min={0}
          {...(max ? { max: getMaxHoursAndMinutes().maxHours } : {})}
          maxLength={getMaxHourLength()}
          value={getHoursString() ?? ''}
          onChange={handleHoursChanged}
          onBlur={handleBlur}
          onFocus={onFocus}
          onKeyDown={onKeyDown}
          type="text"
          inputMode="numeric"
          className={theme['duration-input-numeric-input']}
          autoFocus={autoFocus}
          textAlignRight={textAlignRight}
          size={size}
        />
        <NumericInput
          placeholder="00"
          step={MINUTES_STEP}
          {...(!value?.hours ? { min: 0 } : {})}
          {...(isMaximumMinutesLimited() ? { max: getMaxHoursAndMinutes().maxMinutes } : {})}
          maxLength={getMaxMinuteLength()}
          value={getMinutesString() ?? ''}
          onChange={handleMinutesChange}
          onBlur={handleBlur}
          onFocus={onFocus}
          onKeyDown={onKeyDown}
          type="text"
          inputMode="numeric"
          className={theme['duration-input-numeric-input']}
          textAlignRight={textAlignRight}
          size={size}
        />
      </Box>
    );
  },
);

DurationInput.displayName = 'DurationInput';

export default DurationInput;
