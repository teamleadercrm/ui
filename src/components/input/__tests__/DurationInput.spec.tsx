import React from 'react';
import { render } from '@testing-library/react';
import DurationInput from '../DurationInput';

describe('Component - DurationInput', () => {
  it('renders', async () => {
    const { asFragment } = render(<DurationInput value={{ hours: 3, minutes: 45 }} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
