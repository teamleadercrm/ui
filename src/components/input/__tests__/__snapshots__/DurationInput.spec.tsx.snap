// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - DurationInput renders 1`] = `
<DocumentFragment>
  <div
    class="box align-items-center display-flex duration-input-numeric-container"
    data-teamleader-ui="duration-input"
  >
    <div
      class="box wrapper duration-input-numeric-input"
    >
      <div
        class="input-wrapper"
      >
        <div
          class="input-inner-wrapper"
        >
          <input
            class="input is-medium"
            data-teamleader-ui="numeric-input"
            inputmode="numeric"
            max="9007199254740991"
            min="0"
            placeholder="00"
            step="1"
            stepper="suffix"
            type="text"
            value="03"
          />
          <div
            class="suffix-wrapper"
          >
            <div
              class="stepper"
            >
              <button
                class="box align-items-center display-flex neutral darkest stepper-up"
                data-teamleader-ui="icon"
                tabindex="-1"
                type="button"
              >
                <svg
                  fill="currentColor"
                  height="14"
                  preserveAspectRatio="xMidYMid meet"
                  style="opacity: 0.84; vertical-align: middle;"
                  viewBox="0 0 14 14"
                  width="14"
                >
                  <path
                    clip-rule="evenodd"
                    d="M10.536 7.916L7 4.38 3.465 7.916A1 1 0 004.879 9.33L7 7.208 9.121 9.33a1 1 0 101.415-1.414z"
                    fill="currentColor"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
              <button
                class="box align-items-center display-flex neutral darkest stepper-down"
                data-teamleader-ui="icon"
                tabindex="-1"
                type="button"
              >
                <svg
                  fill="currentColor"
                  height="14"
                  preserveAspectRatio="xMidYMid meet"
                  style="opacity: 0.84; vertical-align: middle;"
                  viewBox="0 0 14 14"
                  width="14"
                >
                  <path
                    clip-rule="evenodd"
                    d="M3.465 6.087L7 9.623l3.536-3.536A1 1 0 009.12 4.673L7 6.794 4.878 4.673a1 1 0 10-1.414 1.414z"
                    fill="currentColor"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="box wrapper duration-input-numeric-input"
    >
      <div
        class="input-wrapper"
      >
        <div
          class="input-inner-wrapper"
        >
          <input
            class="input is-medium"
            data-teamleader-ui="numeric-input"
            inputmode="numeric"
            max="9007199254740991"
            maxlength="2"
            min="-9007199254740991"
            placeholder="00"
            step="15"
            stepper="suffix"
            type="text"
            value="45"
          />
          <div
            class="suffix-wrapper"
          >
            <div
              class="stepper"
            >
              <button
                class="box align-items-center display-flex neutral darkest stepper-up"
                data-teamleader-ui="icon"
                tabindex="-1"
                type="button"
              >
                <svg
                  fill="currentColor"
                  height="14"
                  preserveAspectRatio="xMidYMid meet"
                  style="opacity: 0.84; vertical-align: middle;"
                  viewBox="0 0 14 14"
                  width="14"
                >
                  <path
                    clip-rule="evenodd"
                    d="M10.536 7.916L7 4.38 3.465 7.916A1 1 0 004.879 9.33L7 7.208 9.121 9.33a1 1 0 101.415-1.414z"
                    fill="currentColor"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
              <button
                class="box align-items-center display-flex neutral darkest stepper-down"
                data-teamleader-ui="icon"
                tabindex="-1"
                type="button"
              >
                <svg
                  fill="currentColor"
                  height="14"
                  preserveAspectRatio="xMidYMid meet"
                  style="opacity: 0.84; vertical-align: middle;"
                  viewBox="0 0 14 14"
                  width="14"
                >
                  <path
                    clip-rule="evenodd"
                    d="M3.465 6.087L7 9.623l3.536-3.536A1 1 0 009.12 4.673L7 6.794 4.878 4.673a1 1 0 10-1.414 1.414z"
                    fill="currentColor"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
